src/python/grpcio_tests/tests_aio/unit/interceptor_test.py

```diff
@@ -0,0 +1,340 @@
+                '/grpc.testing.TestService/UnaryCall',
+                request_serializer=messages_pb2.SimpleRequest.SerializeToString,
+                response_deserializer=messages_pb2.SimpleResponse.FromString)
+            call = multicallable(messages_pb2.SimpleRequest())
+
+            response_read = False
+
+            async def read_response():
+                nonlocal response_read
+                await call
+                response_read = True
+
+            task = asyncio.ensure_future(read_response())
+            await asyncio.sleep(0)
+            self.assertFalse(response_read)
+
+            interceptor_lock.set()
+            await task
+            self.assertTrue(response_read)
+
+    async def test_cancel_before_rpc(self):
+
+        interceptor_reached = asyncio.Event()
+
+        class Interceptor(aio.UnaryUnaryClientInterceptor):
+
+            async def intercept_unary_unary(self, continuation,
+                                            client_call_details, request):
+                interceptor_reached.set()
+                await asyncio.sleep(0)
+
+                # This line should never be reached
+                raise Exception()
+
+        server_target, _ = await start_test_server()  # pylint: disable=unused-variable
+
+        async with aio.insecure_channel(
+                server_target, interceptors=[Interceptor()]) as channel:
+
+            multicallable = channel.unary_unary(
+                '/grpc.testing.TestService/UnaryCall',
+                request_serializer=messages_pb2.SimpleRequest.SerializeToString,
+                response_deserializer=messages_pb2.SimpleResponse.FromString)
+            call = multicallable(messages_pb2.SimpleRequest())
+
+            self.assertFalse(call.cancelled())
+            self.assertFalse(call.done())
+
+            await interceptor_reached.wait()
+            self.assertTrue(call.cancel())
+            self.assertFalse(call.cancel())
+
+            with self.assertRaises(asyncio.CancelledError):
+                await call
+
+            self.assertTrue(call.cancelled())
+            self.assertTrue(call.done())
+            self.assertEqual(await call.code(), grpc.StatusCode.CANCELLED)
+            self.assertEqual(
+                await call.details(),
+                'Locally cancelled by application before starting the RPC!')
+            self.assertEqual(await call.initial_metadata(), None)
+            self.assertEqual(await call.trailing_metadata(), None)
+
+    async def test_cancel_after_rpc(self):
+
+        interceptor_reached = asyncio.Event()
+
+        class Interceptor(aio.UnaryUnaryClientInterceptor):
+
+            async def intercept_unary_unary(self, continuation,
+                                            client_call_details, request):
+                call = await continuation(client_call_details, request)
+                await call
+                interceptor_reached.set()
+                await asyncio.sleep(0)
```

Comment:
Lidi and I are wondering if this is potentially racy. Can you run this test a thousand times or so and tell us the flake rate? Try using `--runs_per_test=1000` under Bazel.